### PR TITLE
break dep on clientcmd in pkgs outside pkg oc

### PR DESCRIPTION
**Pull request by juanvallejo**
This patch is a part of pull/17356 - it aims to break all dependencies
between packages outside of `pkg/oc` and the `clientcmd` package.

To achieve this, this patch creates a new package containing only the
functions and objects found in `clientcmd` that are needed by its
dependents outside of the `pkg/oc` subtree. Once this is done, all of
the remaining logic (which should only be used by packages within
`pkg/oc` is moved to `pkg/oc/cli/util/clientcmd` by pull/17356).

This change acknowledges the possibility of having dependents for
`pkg/cmd/util/clientconfig` within the `pkg/oc` subtree.

cc @deads2k @openshift/cli-review @liggitt

### CODE REVIEW

File: pkg/cmd/util/clientconfig/clientconfig.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package clientconfig
```
Comment (deads2k) [inline pkg/cmd/util/clientconfig/clientconfig.go]: not in `pkg/cmd`  I think that should die soon.  `pkg/client/cmd` perhaps.

File: pkg/cmd/server/origin/controller/config.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ func getOpenShiftClientEnvVars(options configapi.MasterConfig) ([]kapi.EnvVar, e
 	if err != nil {
 		return nil, err
 	}
-	return clientcmd.EnvVars(
+	return clientconfig.EnvVars(
```
Comment (deads2k) [inline pkg/cmd/server/origin/controller/config.go]: This appears to be the only caller.  Move it locally.

Comment (deads2k) [reply pkg/cmd/server/origin/controller/config.go]: Actually the entire `getOpenShiftClientEnvVars` needs refactoring.  Queue up an issue to tidy it up.

File: pkg/client/config/loader.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	OpenShiftConfigHomeFileName    = "config"
+	OpenShiftConfigHomeDirFileName = OpenShiftConfigHomeDir + "/" + OpenShiftConfigHomeFileName
+)
+
+var RecommendedHomeFile = path.Join(homedir.HomeDir(), OpenShiftConfigHomeDirFileName)
+
+// currentMigrationRules returns a map that holds the history of recommended home directories used in previous versions.
+// Any future changes to RecommendedHomeFile and related are expected to add a migration rule here, in order to make
+// sure existing config files are migrated to their new locations properly.
+func CurrentMigrationRules() map[string]string {
```
Comment (deads2k) [inline pkg/client/config/loader.go]: do the commands using this care about migration rules?  Which ones are they?

File: pkg/cmd/server/origin/controller/config.go (hunk 2 above)
Comment (juanvallejo) [reply pkg/cmd/server/origin/controller/config.go]: Opened issue to track this: https://github.com/openshift/origin/issues/17397

File: pkg/client/config/loader.go (hunk 3 above)
Comment (juanvallejo) [reply pkg/client/config/loader.go]: Our wrapper command for `NewCmdConfig` [calls cmdconfig.NewOpenShiftClientConfigLoadingRules](https://github.com/openshift/origin/blob/master/pkg/oc/cli/cmd/wrappers.go#L745), which in turn [calls `currentMigrationRules`](https://github.com/openshift/origin/blob/master/pkg/oc/cli/config/loader.go#L45)

**Comment by juanvallejo** [comment]
cc @deads2k

File: pkg/client/cmd/clientcmd.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
+package cmd
```
Comment (deads2k) [inline pkg/client/cmd/clientcmd.go]: is this file a straight move?  If not, highlight what isn't for me.

File: pkg/client/config/loader.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package config
```
Comment (deads2k) [inline pkg/client/config/loader.go]: is this a straight move?

Also, open an issue to discuss removing the auto-migration.  It's been over two years.

File: pkg/cmd/server/admin/create_kubeconfig.go [hunk 6]
```diff
@@ -161,15 +160,15 @@ func (o CreateKubeConfigOptions) CreateKubeConfig() (*clientcmdapi.Config, error
 	}
 
 	// determine all the nicknames
-	clusterNick, err := cliconfig.GetClusterNicknameFromURL(o.APIServerURL)
+	clusterNick, err := config.GetClusterNicknameFromURL(o.APIServerURL)
```
Comment (deads2k) [inline pkg/cmd/server/admin/create_kubeconfig.go]: I don't think I care what this nickname is.  Open an issue to discuss using ugly names that assume an immutable file.

File: pkg/cmd/server/origin/controller/config.go [hunk 7]
```diff
@@ -14,10 +14,28 @@ import (
 
 	configapi "github.com/openshift/origin/pkg/cmd/server/api"
 	"github.com/openshift/origin/pkg/cmd/server/crypto"
-	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
 	"github.com/openshift/origin/pkg/cmd/util/variable"
 )
 
+func envVars(host string, caData []byte, insecure bool, bearerTokenFile string) []kapi.EnvVar {
```
Comment (deads2k) [inline pkg/cmd/server/origin/controller/config.go]: Good.  I hate having "general helpers" that are only used in one spot.  It seems like a way to just push debt to others.

**Comment by deads2k** [comment]
The split needs refinement, but I'll accept ugly to move us forward.  Open the issues, link them here. If the bits I think are moves are actually moves, then this lgtm and you can tag after opening issues.

/approve

File: pkg/client/cmd/clientcmd.go (hunk 4 above)
Comment (juanvallejo) [reply pkg/client/cmd/clientcmd.go]: Yes, the entire file (`pkg/cmd/util/clientcmd/clientcmd.go`) was moved here, except for its `EnvVars` helper, which was moved as an unexported helper in `pkg/cmd/server/origin/controller/config.go`

File: pkg/client/config/loader.go (hunk 5 above)
Comment (juanvallejo) [reply pkg/client/config/loader.go]: Not a straight move.
- [NewOpenShiftClientConfigLoadingRules](https://github.com/openshift/origin/blob/master/pkg/oc/cli/config/loader.go#L45) is left duplicated in the original file, as it is needed by packages within the `pkg/oc` subtree
- [NewPathOptions](https://github.com/openshift/origin/blob/master/pkg/oc/cli/config/loader.go#L61) and [NewPathOptionsWithConfig](https://github.com/openshift/origin/blob/master/pkg/oc/cli/config/loader.go#L65) are entirely left in the original file as they are only used by packages within the `pkg/oc` subtree

Comment (deads2k) [reply pkg/client/config/loader.go]: > Not a straight move.
> 
> NewOpenShiftClientConfigLoadingRules is left duplicated in the original file, as it is needed by packages within the pkg/oc subtree
> NewPathOptions and NewPathOptionsWithConfig are entirely left in the original file as they are only used by packages within the pkg/oc subtree

Ok, that's still fine.

File: pkg/client/config/smart_merge.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
+package config
```
[the remark this answers is not in the document]
Comment (juanvallejo) [reply pkg/client/config/smart_merge.go]: Also not a straight move.

- [CreateConfig](https://github.com/openshift/origin/blob/master/pkg/oc/cli/config/smart_merge.go#L105) helper is used by packages within `pkg/oc`.
- [MergeConfig](https://github.com/openshift/origin/blob/master/pkg/oc/cli/config/smart_merge.go#L156) is used by packages within `pkg/oc`. 
- In that same file (`pkg/oc/cli/config/smart_merge.go`), the following helpers have been unexported, as they are only needed by `MergeConfig` and `CreateConfig`:
  - `GetClusterNicknameFromConfig`
  - `GetUserNicknameFromConfig`
  - `GetContextNicknameFromConfig`
  - `FindExistingContextName`

File: pkg/client/config/loader.go (hunk 5 above)
Comment (juanvallejo) [reply pkg/client/config/loader.go]: > Also, open an issue to discuss removing the auto-migration. It's been over two years.

Tracking this here: https://github.com/openshift/origin/issues/17483

File: pkg/client/config/smart_merge.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (juanvallejo) [reply pkg/client/config/smart_merge.go]: > This smells bad. Open an issue to re-assess naming in node config command.

Tracking this here: https://github.com/openshift/origin/issues/17484

File: pkg/cmd/server/admin/create_kubeconfig.go (hunk 6 above)
Comment (juanvallejo) [reply pkg/cmd/server/admin/create_kubeconfig.go]: > I don't think I care what this nickname is. Open an issue to discuss using ugly names that assume an immutable file.

Tracking this here: https://github.com/openshift/origin/issues/17485

File: pkg/cmd/openshift/openshift.go [hunk 9]
```diff
@@ -147,43 +146,7 @@ func newExperimentalCommand(name, fullName string) *cobra.Command {
 	return experimental
```
Comment (juanvallejo) [inline pkg/cmd/openshift/openshift.go]: These changes will not compile until https://github.com/openshift/origin/pull/17486 (which moves experimental commands into pkg/oc) is merged. This function contains the last use of `pkg/oc/cli/util/clientcmd` outside of the `pkg/oc` subtree.

**Comment by juanvallejo** [comment]
Depends on: https://github.com/openshift/origin/pull/17486

File: pkg/oc/cli/util/clientcmd/clientcmd.go [hunk 10]
```diff
@@ -216,3 +164,27 @@ func (cfg *Config) Clients() (kclientset.Interface, error) {
 
 	return kubeClientset, nil
 }
+
+func DefaultClientConfig(flags *pflag.FlagSet) kclientcmd.ClientConfig {
```
Comment (deads2k) [inline pkg/oc/cli/util/clientcmd/clientcmd.go]: is this a duplicate of something else?

File: pkg/cmd/openshift/openshift.go [hunk 11]
```diff
@@ -99,13 +100,11 @@ func NewCommandOpenShift(name string) *cobra.Command {
...
 	}
 
-	f := clientcmd.New(pflag.NewFlagSet("", pflag.ContinueOnError))
-
 	startAllInOne, _ := start.NewCommandStartAllInOne(name, out, errout)
 	root.AddCommand(startAllInOne)
 	root.AddCommand(newExperimentalCommand("ex", name+" ex"))
 	root.AddCommand(newCompletionCommand("completion", name+" completion"))
-	root.AddCommand(cmd.NewCmdVersion(name, f, out, cmd.VersionOptions{PrintEtcdVersion: true, IsServer: true}))
+	root.AddCommand(cmdversion.NewCmdVersion(name, version.GetWithEtcdVersion(), os.Stdout))
```
Comment (deads2k) [inline pkg/cmd/openshift/openshift.go]: show me the output of `openshift version` before and after please.

File: pkg/oc/cli/util/clientcmd/factory_client_access.go [hunk 12]
```diff
@@ -63,7 +62,7 @@ func NewClientAccessFactory(optionalClientConfig kclientcmd.ClientConfig) Client
 	clientConfig := optionalClientConfig
 	if optionalClientConfig == nil {
 		// TODO: there should be two client configs, one for OpenShift, and one for Kubernetes
-		clientConfig = osclientcmd.DefaultClientConfig(flags)
```
Comment (deads2k) [inline pkg/oc/cli/util/clientcmd/factory_client_access.go]: the dependency in this direction is ok.  You have a lot of them and oc has to vendor origin anyway.

File: pkg/version/version.go [hunk 13]
```diff
@@ -32,6 +34,11 @@ type Info struct {
 	BuildDate  string `json:"buildDate"`
 }
 
+type InfoWithEtcd struct {
```
Comment (deads2k) [inline pkg/version/version.go]: I don't think we want to do this.  How about we build a separate version command for `openshift`.  It really is special.  It has multiple versions to show.

File: pkg/oc/cli/util/clientcmd/clientcmd.go (hunk 10 above)
Comment (juanvallejo) [reply pkg/oc/cli/util/clientcmd/clientcmd.go]: Yes, this function is also present in `pkg/client/cmd/clientcmd.go`: https://github.com/openshift/origin/pull/17357/files#diff-dc9ab0f48687544bfe9cc38fab5ecac3R207

File: pkg/cmd/openshift/openshift.go (hunk 11 above)
Comment (juanvallejo) [reply pkg/cmd/openshift/openshift.go]: **Before**
```
$ openshift version
openshift v3.8.0-alpha.0+da1cca8-325-dirty
kubernetes v1.8.1+0d5291c
etcd 3.2.8
```

**After**
```
$ openshift version
openshift v3.8.0-alpha.0+9d4c131-327-dirty
etcd 3.2.8
```

Per https://github.com/openshift/origin/pull/17357#discussion_r153891196, will make a new version command for `openshift` to address this discrepancy

**Comment by juanvallejo** [comment]
Added separate version command for `openshift`:

```
$ openshift version
openshift v3.8.0-alpha.0+9d4c131-327-dirty
kubernetes v1.8.1+0d5291c
etcd 3.2.8
```

**Comment by deads2k** [comment]
@juanvallejo failure to build

**Comment by juanvallejo** [comment]
@deads2k this won't build until https://github.com/openshift/origin/pull/17486 merges and I rebase

**Comment by deads2k** [comment]
@juanvallejo  start writing up flake issues for the different e2e issues.

**Comment by juanvallejo** [comment]
extended_clusterup flaked on https://github.com/openshift/origin/issues/17548

**Comment by juanvallejo** [comment]
crio test flaked on https://github.com/openshift/origin/issues/14898 - error talking to yum

**Pull request merged**